bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/ExtensibleTrustManager.java

```diff
@@ -0,0 +1,180 @@
+            defaultTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        } else {
+            linkedTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        }
+    }
+
+    private X509ExtendedTrustManager getLinkedTrustMananger(X509Certificate[] chain) {
+        try {
+            String commonName = getCommonName(chain[0]);
+
+            X509ExtendedTrustManager trustManager = linkedTrustManager.get(commonName);
+
+            if (trustManager != null) {
+                logger.trace("Found trustManager by common name: {}", commonName);
+                return trustManager;
+            }
+
+            logger.trace("Searching trustManager by Subject Alternative Names: {}",
+                    chain[0].getSubjectAlternativeNames());
+            // @formatter:off
+            return chain[0].getSubjectAlternativeNames().stream()
+                    .map(e -> e.get(1))
+                    .map(Object::toString)
+                    .map(linkedTrustManager::get)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElse(null);
+            // @formatter:on
+        } catch (CertificateParsingException e) {
+            throw new IllegalStateException("Problem while parsing certificate", e);
+        }
+    }
+
+    private String getCommonName(X509Certificate x509Certificate) {
+        String dn = x509Certificate.getSubjectX500Principal().getName(X500Principal.RFC2253);
+        for (String group : dn.split(",")) {
+            if (group.contains("CN=")) {
+                return group.trim().replace("CN=", "");
+            }
+        }
+        throw new IllegalStateException("No Common Name found");
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        logger.trace("adding implementation of TlsCertificateProvider for {} to ExtensibleTrustManager",
+                tlsCertificateProvider.getHostName());
+        linkedTrustManager.put(tlsCertificateProvider.getHostName(),
+                new TlsCertificateTrustManagerAdapter(tlsCertificateProvider).getTrustManager());
+        logger.trace("Successfully add implementation of TlsCertificateProvider for {} to ExtensibleTrustManager",
+                tlsCertificateProvider.getHostName());
+    }
+
+    protected void removeTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        // TODO this might remove the wrong one, should we do a better check here?
+        linkedTrustManager.remove(tlsCertificateProvider.getHostName());
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addTlsTrustManagerProvider(TlsTrustManagerProvider tlsTrustManagerProvider) {
+        // TODO we could potentially overwrite existing implementation do we want to guard against that?
```

Comment:
@kaikreuzer wdyt should there by any checks / warnings in case to Components implement a `tls..Provider` for the same hostname? Or should it even support it?

Reply:
I would log a warning and ignore any additional trust managers for urls for which we already have one. Just to prevent that during runtime someone might try to take over control of a specific url.

Reply:
Good idea 👍

Reply:
Hm, if multiple trust managers for the same hostname are used, shouldn't we drop ALL the trust managers and log a big fat warning?
I assume we don't know which one is the correct one.
Perhaps the one that tries to take over the specific URL is loaded before the correct one after the next restart.

Reply:
It would feel wrong that a fully working system suddenly stops working, just because you have dynamically added another independent bundle that happens to use the same url. Imho a warning in that case is good enough and that's also how we proceed in all other cases where we expect a single service, but (dynamically) might get multiple.

Reply:
Hm, for me (personally) it feels more wrong to just write a log message? Who reads log messages if he think all is working fine?. If e.g. the windows defender identifies a potential security attack and write a log message I assume no one gets noticed ever.

Perhaps the situation is different here. I don't know.

If a binding communicates with a remote server and transfer security related things (isn't the most "home" automation more or less some very personal information) shouldn't we ensure that the certificate checking of a peer cannot be levered out easily? If an attacker can inject a `TlsTrustManagerProvider` and ensure it is loaded earlier then the correct one (for a specific hostname) shouldn't we be very noisy that there is something wrong (two providers for the same hostname)?

Sure, if someone can inject a HTTP client factory that is used instead of the framework provided one it is a more bad situation...
So perhaps a warning is enough as we cannot really identify such an attack...